src/shuffler-codereview.c

```diff
@@ -0,0 +1,1228 @@
+
+#include "shuffle_trainers.h"
+
+#ifdef __wasm__
+#define MYLOG(f_, ...) printf("GBA: "); printf((f_), ##__VA_ARGS__); printf("\n")
+#else
+#define MYLOG(f_, ...) MgbaPrintf(MGBA_LOG_INFO, (f_), ##__VA_ARGS__)
+#endif
+
+static const u8 unknown_string[] = _("unknown");
+
+extern struct Evolution gEvolutionTable[][EVOS_PER_MON];
+extern u32 GetTotalBaseStat(u32 species);
+
+EWRAM_DATA u32 seed;
+EWRAM_DATA tinymt32_t masterSeed;
+EWRAM_DATA u16 realStarterMon[TOTAL_STARTERS];
+
+EWRAM_DATA struct MapEvents AdjustedMapEvents;
+EWRAM_DATA struct ObjectEventTemplate AdjustedTemplates[MAX_OBJECTS];
+EWRAM_DATA u16 CurrentAdjustedRoom = 0xFFFF;
+EWRAM_DATA u8 CurrentAdjustedRoomIndex;
+
+// a bitfield used to keep track of which overworld objects are being used.
+// useful for objects with partners when determining if the partner has
+// already been created or not.
+EWRAM_DATA u16 AdjustedObjectsDirty = 0;
+// normally each object has its own flag, but some objects share their flag
+// with their partner. if an object is determined to be the second in the
+// pair, it gets marked as such via this bitfield.
+EWRAM_DATA u16 RedirectedTrainerFlags = 0;
+// randomized overworld objects. normally overworld objects are read directly
+// from ROM data, but we generate them when the player enters the room, store
+// them here, and redirect calls the game makes to this object.
+EWRAM_DATA union ShuffledObject AdjustedObjects[MAX_OBJECTS];
+
+EWRAM_DATA struct WarpData realWarps[ROOMS_PER_WAVE] = {};
+// the current distance in rooms that the player is from the starting room.
+// mostly used for determining level of wild and npc pokemon, but also for
+// things like room selection.
+EWRAM_DATA u8 distance = 0;
+EWRAM_DATA u16 lastRoom = 0;
+
+EWRAM_DATA u8 NextGymType = 0xFF;
+EWRAM_DATA u8 NextGymMinorType = 0xFF;
+
+// I'm planning on eventually pulling some of these helper functions into their own file.
+void FillArrayWithOrderedIndicies(int* arr, int length) {
+    for (int i = 0; i < length; i++) {
+        arr[i] = i;
+    }
+}
+
+void ShuffleArray(int* arr, int length, tinymt32_t* rng) {
+    for (int i = length - 1; i > 0; i--) {
+        int r = tinymt32_generate_uint32(rng) % (i + 1);
+        int temp;
+        SWAP(arr[i], arr[r], temp);
+    }
+}
+
+u16 GetRoomId(u8 mapGroup, u8 mapNum) {
+    return (mapGroup << 8) | mapNum;
+}
+
+void AddInitialBagItems(tinymt32_t* rng) {
+    for (int i = 0; i < 5; i++) {
+        AddBagItem(ITEM_TM01 + (tinymt32_generate_uint32(rng) % 50), 1);
+    }
+
+    AddBagItem(ITEM_SITRUS_BERRY, 5);
+    AddBagItem(ITEM_LEPPA_BERRY, 5);
+
+    AddBagItem(ITEM_POKE_BALL, 8);
+    AddBagItem(ITEM_GREAT_BALL, 4);
+    AddBagItem(ITEM_ULTRA_BALL, 2);
+    AddBagItem(ITEM_MASTER_BALL, 1);
+}
+
+void ShuffleStarters(tinymt32_t* rng) {
+    int nextIndex = 0;
+    // generate the list of potential starting pokemon.
+    for (int i = POSSIBLE_STARTERS - 1; i >= POSSIBLE_STARTERS - TOTAL_STARTERS; i--) {
```

Comment:
I think this is the O(N^2) we were talking about last weekend? I think it might be worse than that – possibly not terminating? We should talk again, there's probably a better way to do this.